metricbeat/module/windows/wmi/utils.go

```diff
@@ -0,0 +1,200 @@
+	// Documentation: https://learn.microsoft.com/en-us/windows/win32/wmisdk/swbemobject-properties-
+	rawResult, err := oleutil.GetProperty(instance.GetIDispatch(), "Properties_")
+	if err != nil {
+		return nil, err
+	}
+
+	// SWbemObjectEx.Properties_ returns
+	// an SWbemPropertySet object that contains the collection
+	// of properties for the c class
+	sWbemObjectExAsIDispatch := rawResult.ToIDispatch()
+	defer func() {
+		if cerr := rawResult.Clear(); cerr != nil {
+			logger.Debugf("failed to release connection: %w", err)
+		}
+	}()
+
+	// Get the property
+	sWbemProperty, err := oleutil.CallMethod(sWbemObjectExAsIDispatch, "Item", propertyName)
+	if err != nil {
+		return nil, err
+	}
+
+	return sWbemProperty.ToIDispatch(), nil
+}
+
+// Given an instance and a property Name, it returns the appropriate conversion function
+func GetConvertFunction(instance *wmi.WmiInstance, propertyName string, logger *logp.Logger) (WmiStringConversionFunction, error) {
+	rawProperty, err := getProperty(instance, propertyName, logger)
+	if err != nil {
+		return nil, err
+	}
+	propType := getPropertyType(rawProperty)
+
+	var f WmiStringConversionFunction
+
+	switch propType {
+	case base.WbemCimtypeDatetime:
+		f = ConvertDatetime
+	case base.WbemCimtypeUint64:
+		f = ConvertUint64
+	case base.WbemCimtypeSint64:
+		f = ConvertSint64
+	default: // For all other types we return the identity function
+		f = ConvertString
+	}
+	return f, err
+}
+
+// Utilities related to Warning Threshold
+
+// Define an interface to allow unit-testing long-running queries
+// *wmi.wmiSession is an implementation of this interface
+type WmiQueryInterface interface {
+	QueryInstances(query string) ([]*wmi.WmiInstance, error)
+}
+
+// Wrapper of the session.QueryInstances function that execute a query for at most a timeout
+// after which we stop actively waiting.
+// Note that the underlying query will continue to run, until the query completes or the WMI Arbitrator stops the query
+// https://learn.microsoft.com/en-us/troubleshoot/windows-server/system-management-components/new-wmi-arbitrator-behavior-in-windows-server
+func ExecuteGuardedQueryInstances(session WmiQueryInterface, query string, timeout time.Duration, logger *logp.Logger) ([]*wmi.WmiInstance, error) {
+	var rows []*wmi.WmiInstance
+	var err error
+	done := make(chan error)
+	timedout := false
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	go func() {
+		start_time := time.Now()
+		rows, err = session.QueryInstances(query)
+		if !timedout {
+			done <- err
+		} else {
+			timeSince := time.Since(start_time)
+			baseMessage := fmt.Sprintf("The query '%s' that exceeded the warning threshold terminated after %s", query, timeSince)
+			var tailMessage string
+			// We eventually fetched the documents, let us free them
+			if err == nil {
+				tailMessage = "successfully. The result will be ignored"
```

Comment:
seems a shame to throw away the result. perhaps we can revisit this in a later PR?

Reply:
I agree, once it's closed, I will open a PR for this